### PR TITLE
 Clean up method body encoding APIs

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera Preliminary, I have not run any tests at all :)

**Comment by tmat** [comment]
@nguerrera Please, review.

File: src/System.Reflection.Metadata/src/Resources/Strings.resx [hunk 1]
```diff
@@ -456,4 +459,13 @@
   <data name="UnsupportedFormatVersion" xml:space="preserve">
     <value>Unsupported format version: {0}</value>
   </data>
+  <data name="DistanceBetweenInstructionAndLabelTooBig" xml:space="preserve">
+    <value>The distance between the instruction {0} (offset {1}) and the target label is doesn't fit the operand size: {2}</value>
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/Resources/Strings.resx]: typo: "is doesn't fit" should be "doesn't fit"

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/ExceptionRegionEncoder.cs [hunk 2]
```diff
@@ -2,120 +2,145 @@
...
+
+        internal static bool IsSmallExceptionRegionFromBounds(int startOffset, int endOffset) => 
+            IsSmallExceptionRegion(startOffset, endOffset - startOffset);
+
+        internal static int GetExceptionTableSize(int exceptionRegionCount, bool isSmallFormat) => 
+            TableHeaderSize + exceptionRegionCount * (isSmallFormat ? SmallRegionSize : FatRegionSize);
+
+        internal static bool IsExceptionRegionCountInBounds(int exceptionRegionCount) => 
+            TableHeaderSize + (long)unchecked((uint)exceptionRegionCount) * FatRegionSize <= 0xffffff; 
+
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/ExceptionRegionEncoder.cs]: I suggest (byte|ushort|uint).MaxValue above. Easier to read than counting how 'f's there are IMHO.

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/ExceptionRegionEncoder.cs]: Ah, of course of them (0xffffff) is none of the above. Separate constant for that would be nice.

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/ExceptionRegionEncoder.cs]: OK.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                Throw.ArgumentOutOfRange(nameof(exceptionRegionCount));
+            }
+
+            int bodyOffset = SerializeHeader(codeSize, (ushort)maxStack, exceptionRegionCount, attributes, localVariablesSignature);
+            var instructions = Builder.ReserveBytes(codeSize);
+
+            var regionEncoder = (exceptionRegionCount > 0) ? 
+                ExceptionRegionEncoder.SerializeTableHeader(Builder, exceptionRegionCount, hasFatExceptionRegions) : default(ExceptionRegionEncoder);
+
+            // return a tuple
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: This is an odd comment.

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: MethodBody would be just a tuple if we could use tuples with named items.

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: Is this a TODO, then? I thought we settled on proper structs apart from the special case where the order is relevant right? If this is the proposed final shape, then we don't need this comment.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+            return new MethodBody(bodyOffset, instructions, regionEncoder);
+        }
+
+        public struct MethodBody
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: This being nested makes MethodBodyStreamEncoder.MethodBody long and repetitive. I know we already struggled for MethodBodyBlock name. OK to leave to API review.

File: src/System.Reflection.Metadata/tests/Metadata/Ecma335/Encoding/ControlFlowBUilderTEsts.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+using Xunit;
+
+namespace System.Reflection.Metadata.Ecma335.Tests
+{
+    public class ControlFlowBuilderTests
+    {
+        [Fact]
+        public void F()
+        {
+            // TODO
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/tests/Metadata/Ecma335/Encoding/ControlFlowBUilderTEsts.cs]: Nice test. ;)

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs [hunk 6]
```diff
@@ -259,22 +263,14 @@ private static MethodDefinitionHandle BasicValidationEmit(MetadataBuilder metada
...
-            var body = methodBodies.AddMethodBody(exceptionRegionCount: 1);
-            var eh = body.WriteInstructions(codeBuilder, branchBuilder, out mainBodyOffset);
-            eh.StartRegions();
-            eh.AddFinally(tryOffset, handlerOffset - tryOffset, handlerOffset, handlerEnd - handlerOffset);
-            eh.EndRegions();
-
+            int mainBodyOffset = methodBodyStream.AddMethodBody(il);
             codeBuilder.Clear();
-            branchBuilder.Clear();
+            flowBuilder.Clear();
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: The exception handling with labels via ControlFlowBuilder is really nice vs. what you had to do before! :)

**Comment by nguerrera** [comment]
:+1:

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs (hunk 4 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: That's the problem with not having ValueTuples - I don't want to pollute the namespace with this struct. I don't think anybody will need to actually use the type. Typically the usage would be either ```var (offset, instr, ex) = AddMethodBody()``` in C# 7 or ```var body = AddMethodBody()```. I'd also leave it to the API review though.

File: src/System.Reflection.Metadata/tests/Metadata/Ecma335/Encoding/ControlFlowBUilderTEsts.cs (hunk 5 above)
Comment (tmat) [inline src/System.Reflection.Metadata/tests/Metadata/Ecma335/Encoding/ControlFlowBUilderTEsts.cs]: It passes! ;)

File: src/System.Reflection.Metadata/src/Resources/Strings.resx (hunk 1 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/Resources/Strings.resx]: Fixed

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs (hunk 3 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs]: Removed

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs (hunk 6 above)
Comment (tmat) [inline src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: Glad you like it :)

**Comment by tmat** [comment]
test ubuntu14.04_release_prtest please

**Pull request merged**